void_migration/void_migration.py

```diff
@@ -466,6 +516,156 @@ def close_voids(u, v, s):
+    this is done in the time_march function 
+    '''
+    
+    s_ms = 0
+    if p.gsd_mode == 'mono':
+
+        colsm = []
+        for i in range(p.no_of_cycles):
+            if i % 2 == 0:
+                colsm.append(change_s_ms[0])
+            else:
+                colsm.append(change_s_ms[1])
+
+    res = [sub['t_empty'] for sub in op_arr]  #Just take the t_empty values from the op_arr dictionary
+    
+    yj = []
+    for j in range(len(res)):
+        if t < int(np.ceil(res[j]/p.dt)):    
+            yj.append(op_arr[j])
+    
+    if p.gsd_mode == 'mono':
+        col_index = colsm[int(len(op_arr) - len(yj))]
+
+
+    if t <= int(np.ceil(yj[0].get('t_fill')/p.dt)):   ##To always pick the time belonging to the first element in the array 
+        if p.gsd_mode == 'mono':
+            p.s_m = col_index
+            p.add_voids = 'place_on_top' #'pour_base'
+        else:
+            p.add_voids = 'place_on_top'
+
+    elif int(np.ceil(yj[0].get('t_fill')/p.dt)) < t <= int(np.ceil(yj[0].get('t_settle')/p.dt)):
+        p.add_voids = 'None'
+
+    elif int(np.ceil(yj[0].get('t_settle')/p.dt)) < t < int(np.ceil(yj[0].get('t_empty')/p.dt)-1):
+        p.add_voids = "central_outlet"
+        p.save_outlet = True
+
+    return p,change_s_ms
+
+
+def cals_in_cd(p):
+    '''
+    Calculates the time required for filling, settling and emptying for each cycle and a desired number of cycles
+    M_total : Total mass for a given solid fraction
+    Mass_in_per_u_t : Mass going in per unit time
+    Mass_out_per_u_t : Mass coming out of silo per unit time 
+    T_in : filling time calculated based on M_total, Mass_in_per_u_t and p.dt
+    T_settle : settling time 
+    T_out : emptying time calculated based on M_in, Mass_out_per_u_t and p.dt
+    op_arr : holds the values for t_fill, t_settle and t_empty for each cycle and for a desired number of cycles  
+
+    '''
+    if p.gsd_mode == 'mono':
+        v_sphere = (4/3) * np.pi * ((p.s_m + p.s_M)/2/2)**3   ## m^3
+    elif p.gsd_mode == 'bi':
+        v_sphere = (((4/3)*np.pi*(p.s_m/2)**3) + ((4/3)*np.pi*(p.s_M/2)**3)) / 2
+    else:
+        v_sphere = (4/3) * np.pi * ((p.s_m + p.s_M)/2/2)**3   ## m^3
+
+    rho_p = 3950    ## kg/m^3
+
+    # Total mass
+    M_total = p.nx * p.ny * p.nm * v_sphere * rho_p * p.nu_cs
+
+    free_par = 0.17             ## do not know why this. How to decide this?
+
+    # Mass per unit time
+    Mass_in_per_u_t = (p.half_width * 2 + 1) * p.nm * p.fill_ratio * v_sphere * rho_p #* p.nu_cs
+    Mass_out_per_u_t = (p.half_width * 2 + 1) * p.nm * free_par * v_sphere * rho_p #* p.nu_cs
+
+    # Mass in
+    Mass_in = [M_total/2,M_total/2,M_total/2]
+
+    # Mass out
+    Mass_out = [Mass_in[0]/1.1,Mass_in[1]/1.1,Mass_in[2]/1.1]
+
+    # Time in 
+    T_in = [int((Mass_in[0]/Mass_in_per_u_t)*p.dt),int((Mass_in[1]/Mass_in_per_u_t)*p.dt),int((Mass_in[2]/Mass_in_per_u_t)*p.dt)]
+
+    # Time settle 
+    T_settle = [3, 3, 3]
```

Comment:
this should be in the parameters file or at least a default value in defaults.json5